ee/hogai/utils/stream_processor.py

```diff
@@ -16,145 +16,188 @@
+                return [result]
 
-            # Register any tool calls for later parent chain lookups
-            self._register_tool_calls(message)
-            result = self._handle_message(message, cast(MaxNodeName, node_name))
-            return (
-                result if result is not None else AssistantGenerationStatusEvent(type=AssistantGenerationStatusType.ACK)
-            )
+        if isinstance(action, UpdateAction) and (update_event := self._handle_update_message(event, action)):
+            return [update_event]
 
-    def _find_parent_ids(self, message: AssistantMessage) -> tuple[str | None, str | None]:
-        """
-        Walk up the message chain to find the root parent's message_id and tool_call_id.
+        return None
 
-        Returns (root_message_id, root_tool_call_id) for the root message in the chain.
-        Includes cycle detection and max depth protection.
+    def process_langgraph_update(self, event: LangGraphUpdateEvent) -> list[AssistantResultUnion] | None:
+        """
+        Process a LangGraph update event.
         """
-        root_tool_call_id = message.parent_tool_call_id
-        if root_tool_call_id is None:
-            return message.id, None
-
-        root_message_id = None
-        visited: set[str] = set()
-
-        while root_tool_call_id is not None:
-            if root_tool_call_id in visited:
-                # Cycle detected, we skip this message
-                return None, None
-
-            visited.add(root_tool_call_id)
-            parent_message = self._tool_call_id_to_message.get(root_tool_call_id)
-            if parent_message is None:
-                # The parent message is not registered, we skip this message as it could come
-                # from a sub-nested graph invoked directly by a contextual tool.
-                return None, None
-
-            next_parent_tool_call_id = parent_message.parent_tool_call_id
-            root_message_id = parent_message.id
-            if next_parent_tool_call_id is None:
-                return root_message_id, root_tool_call_id
-            root_tool_call_id = next_parent_tool_call_id
-        raise ValueError("Should not reach here")
-
-    def _register_tool_calls(self, message: AssistantMessageUnion) -> None:
-        """Register any tool calls in the message for later lookup."""
-        if isinstance(message, AssistantMessage) and message.tool_calls is not None:
-            for tool_call in message.tool_calls:
-                self._tool_call_id_to_message[tool_call.id] = message
+        if is_message_update(event.update):
+            # Convert the message chunk update to a dispatcher event to prepare for a bright future without LangGraph
+            maybe_message_chunk, state = event.update[1]
+            if not isinstance(maybe_message_chunk, AIMessageChunk):
+                return None
+            action = AssistantDispatcherEvent(
+                action=MessageChunkAction(message=maybe_message_chunk),
```

Comment:
Question: I don't see `node_path`, so `_is_message_from_nested_node_or_graph` won't determine message routing, so I assume it will be always treated as a root-level message? I don't fully understand how it works, so would appreciate a brief explanation 😄